### DISCUSSION TITLE
SQLAlchemy: how to specify WHERE conditions with a string?

### DISCUSSION THREAD

**Discussion by moonman239** [sqlalchemy/sqlalchemy#8239] [Usage Questions] [answered]
I am building an API where the client is to be allowed to retrieve and to filter a set of records.
I know that I can hard-code a filter myself, ex: ```select(Record).where(Record.id == 5)```

I want to know if SQLAlchemy or SQL has a good way to specify the WHERE condition with a string, ex:

```
filter = 'Record.id == 5'
select(Record).where(filter)
```

I could use something like
```
filter = 'Record.id == 5'
query = text('SELECT * FROM RecordTable WHERE " + filter)
```

but that opens my program up to SQL injections.

Am I asking for something that doesn't exist?

Please help and thanks.

**Comment by CaselIT** [comment, marked as answer]
Hi,

short answer is no, if you plan to accept any str as filter and run it you need to make sure that only trusted clients can access that api.
otherwise if the scope is only to support where it should not be too difficult to pass the condition in some format that could be generated on the server. Something like

```py
filters = [['col', 'eq', 123], ['col2', 'lt', 456]]
q = select(Record)
for c, op, value in filters:
  q = q.where(getattr(sa.sql.operators, op)(sa.column(c), value)
```

**Reply by moonman239** [reply]
That looks nice, but I need a solution that supports queries like
``` (col3 eq True) and ((col4 lt 789) or (col5 gt 180)```

**Reply by moonman239** [reply]
though I suppose I could add more dimensions, ex:
```filters = [[['col3','eq','True'],'and',[['col4','lt',789'],'or',['col5','gt',180']]]]```

**Reply by CaselIT** [reply]
indeed not. that was just an example though. a recursive structure with the type of conjunction can be used.
otherwise if you are ok with minimizing the condition in two levels that could be simplified by assuming a two level ORs of ANDs

**Reply by moonman239** [reply]
@CaselIT I expanded your solution to allow "and"'s and "or"'s:
```py
def filter_operation(orm_object, filter: str):
    # if contains parentheses, evaluate them first and then 
    # basic operations
    if (len(filter.split(" ")) == 3):
        filter = re.sub(r"[\(|\)]", "", filter)
        operation = filter.split(" ")
        column = getattr(orm_object, operation[0])
        f = getattr(sqlalchemy.sql.operators, operation[1])
        value = operation[2]
        # value can either be number, datetime, or string
        try:
            value = int(value)
        except ValueError as ve:
            # number is not int
            try:
                value = parser.isoparse(value)
            except ValueError as ve:
                print(
                    "value is not a datetime in ISO 8601 format. If it is supposed to be a datetime, it's not properly formatted."
                )
        print("evaluating expression {0}({1},{2}".format(f, column, value))
        return f(column, value)
    else:
        # this function contains "and" and/or "or" operations.
        try:
            and_index = filter.index(" and ")
            anded_operations = [
                filter_operation(orm_object, " ".join(substr.split(" ")[0:3]))
                for substr in filter.split(" and ")
            ]
            return and_(*tuple(anded_operations))
        except ValueError as e:
            print("no 'and' operations")
        try:
            or_index = filter.index(" or ")
            ored_operations = [
                filter_operation(orm_object, " ".join(substr.split(" ")[0:3]))
                for substr in filter.split(" or ")
            ]
            return or_(*tuple(ored_operations))
        except ValueError as e:
            print("no 'or' operation")
```

You use it like so:
```select(ObjectClass).where(filter(ObjectClass,filter_str))```
If filter_str is "col3 eq True and col4 lt 789", this will basically emit
"SELECT ObjectClass.* WHERE col3 = True AND col4 < 789"

**Reply by moonman239** [reply]
Plus, it's theoretically safe from certain SQL injections, because the set of possible inputs is limited, ex. if I try "col3 UNION DELETE * FROM Entries" the function doesn't know what to do with that input, and will simply return "None."

**Reply by moonman239** [reply]
You can even safeguard data from unauthorized access by doing something like
```select(ObjectClass).where(and_(authorization_criteria,filter(ObjectClass,filter_str)))```

**Reply by CaselIT** [reply]
looks fine to me.



> Plus, it's theoretically safe from certain SQL injections, because the set of possible inputs is limited, ex. if I try "col3 UNION DELETE * FROM Entries" the function doesn't know what to do with that input, and will simply return "None."

it should be safe yes, since all the right sides are passed are parameters so the db will see something like `select * from tbl where col = :par` and a parameter like `{'par': 42}`
